services/storage/rpc_service.go

```diff
@@ -31,35 +41,66 @@ func (r *rpcService) Hints(context.Context, *types.Empty) (*HintsResponse, error
 	return nil, errors.New("not implemented")
 }
 
-const (
-	batchSize  = 1000
-	frameCount = 50
-	writeSize  = 64 << 10 // 64k
-)
-
 func (r *rpcService) Read(req *ReadRequest, stream Storage_ReadServer) error {
 	// TODO(sgc): implement frameWriter that handles the details of streaming frames
+	var err error
+	var wire opentracing.SpanContext
+
+	if len(req.Trace) > 0 {
+		wire, err = opentracing.GlobalTracer().Extract(opentracing.TextMap, opentracing.TextMapCarrier(req.Trace))
+		if err != nil {
+			// TODO(sgc): log it?
+		}
+	}
+
+	span := opentracing.StartSpan("storage.read", ext.RPCServerOption(wire))
+	defer span.Finish()
+
+	ext.DBInstance.Set(span, req.Database)
+
+	// TODO(sgc): use yarpc stream.Context() once implemented
+	ctx := context.Background()
+	ctx = opentracing.ContextWithSpan(ctx, span)
+	// TODO(sgc): this should be available via a generic API, such as tsdb.Store
+	ctx = tsm1.NewContextWithMetricsGroup(ctx)
+
+	var agg Aggregate_AggregateType
+	if req.Aggregate != nil {
+		agg = req.Aggregate.Type
+	}
+	pred := PredicateToExprString(req.Predicate)
+	groupKeys := strings.Join(req.Grouping, ",")
+	span.
+		SetTag("predicate", pred).
```

Comment:
This could be quite long, do we want it always attached to the span?

Reply:
are you referring to the `groupKeys` or the `pred` (predicate)?

Reply:
The predicate

Reply:
Actually, how about I use this pattern and only enable if we set an environment variable:

https://github.com/opentracing/opentracing-go#conditionally-capture-a-field-using-lognoop

Reply:
Perhaps we can stick with a `max-tag-length`, that can later be exposed as a configuration option

Reply:
It appears that is already taken care of by the client; `jaeger` for example limits the tag length to 256

Reply:
ok, yeah a limit sounds good, its just informational so if we truncate it, its no big deal. Then we have the data so we can correlate it later.

Reply:
added `truncateString` and limited to 256 also